web/src/server/api/routers/comments.ts

```diff
@@ -0,0 +1,242 @@
+}): Promise<void> => {
+  const { objectId, objectType, projectId } = input;
+  const prismaModel = COMMENT_OBJECT_TYPE_TO_PRISMA_MODEL[objectType];
+
+  if (!prismaModel) {
+    throw new Error(`No prisma model for object type ${objectType}`);
+  }
+
+  const model = ctx.prisma[prismaModel];
+  const object = await model.findFirst({
+    where: {
+      id: objectId,
+      projectId,
+    },
+  });
+
+  if (!object) {
+    throw new Error(
+      `No ${prismaModel} with id ${objectId} in project ${projectId}`,
+    );
+  }
+};
+
+export const commentsRouter = createTRPCRouter({
+  create: protectedProjectProcedure
+    .input(CreateCommentData)
+    .mutation(async ({ input, ctx }) => {
+      try {
+        throwIfNoProjectAccess({
+          session: ctx.session,
+          projectId: input.projectId,
+          scope: "comments:CUD",
+        });
+
+        validateCommentReferenceObject({ ctx, input });
+
+        const comment = await ctx.prisma.comment.create({
+          data: {
+            projectId: input.projectId,
+            content: input.content,
+            objectId: input.objectId,
+            objectType: input.objectType,
+            authorUserId: ctx.session.user.id,
+          },
+        });
+
+        await auditLog({
+          session: ctx.session,
+          resourceType: "comment",
+          resourceId: comment.id,
+          action: "create",
+          after: comment,
+        });
+
+        return comment;
+      } catch (error) {
+        console.error(error);
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "Creating comment failed.",
+        });
+      }
+    }),
+  delete: protectedProjectProcedure
+    .input(DeleteCommentData)
+    .mutation(async ({ input, ctx }) => {
+      try {
+        throwIfNoProjectAccess({
+          session: ctx.session,
+          projectId: input.projectId,
+          scope: "comments:CUD",
+        });
+
+        const comment = await ctx.prisma.comment.findFirst({
+          where: {
+            id: input.id,
+            projectId: input.projectId,
+            objectId: input.objectId,
+            objectType: input.objectType,
+          },
+        });
+        if (!comment) {
+          throw new Error("No comment with this id in this project.");
+        }
+
+        if (comment.authorUserId !== ctx.session.user.id) {
+          throw new Error(
+            "Comment author user id does not match provided user id",
+          );
+        }
+
+        await ctx.prisma.comment.delete({
+          where: {
+            id: comment.id,
+            projectId: comment.projectId,
+            objectId: comment.objectId,
+            objectType: comment.objectType,
+          },
+        });
+
+        await auditLog({
+          session: ctx.session,
+          resourceType: "comment",
+          resourceId: comment.id,
+          action: "delete",
+          before: comment,
+        });
+      } catch (error) {
```

Comment:
do we need try catch block here? hinders us from throwing more specific error messages as they will be catched here

Reply:
With the try ... catch blocks we can ensure we log the errors. I have now slightly adjusted the code to then throw the caught error if it's a TRPC error and only default to internal service error if the error is of an unknown type -- we'd still get the specific error messages in the logs, but have the guarantee that we'd always be throwing TRPC errors. Wdyt?

Reply:
ok, will share try/catch to catch non-trpcerrors in trpc routes